### ISSUE TITLE
Update of PC from within a UC_HOOK_BLOCK callback is ignored

### ISSUE THREAD

**Issue by oTuaz9gu** [unicorn-engine/unicorn#1726]
Hi,

It would seem that updating the program counter from within a callback registered on a block (UC_HOOK_BLOCK) doesn't actually alter the execution flow.

The attached script (see below) attempts to illustrate the issue. It executes the following ARM code in Unicorn:
```s
main:
        mov r0, #0
        b branch_A

branch_A:
        mov r0, #1
        bx lr

branch_B:
        mov r0, #2
        bx lr 
```
1. when the script is called with no arguments, no callback is registered and the program executes normally (exits with r0 = **1**)
2. when called with `--block` and/or `--code` a  UC_HOOK_BLOCK (respectively UC_HOOK_CODE) hook is added. The callback from the hook checks if the current address is branch_A's and, if so, redirects to branch_B.
    Consequently the _expected_ final value for r0 is **2** if any of the script is called with either of these arguments. 
    When run with `--block`, the script yields **1** and yields the expected **2** only when the `--code` argument is used.

This looks eerily similar to #1661 (on ARM64).

---

```python
import argparse
import unicorn
from unicorn.arm_const import *

# main:
#         mov r0, #0
#         b branch_A
#
# branch_A:
#         mov r0, #1
#         bx lr
#
# branch_B:
#         mov r0, #2
#         bx lr 
ARM_CODE = bytes.fromhex('00 00 a0 e3 ff ff ff ea 01 00 a0 e3 1e ff 2f e1 02 00 a0 e3 1e ff 2f e1')

CODE_ADDR = 0x1000
EXIT_ADDR = 0x1100

BA_ADDR = CODE_ADDR + 0x4
BB_ADDR = CODE_ADDR + 0x10


parser = argparse.ArgumentParser()
parser.add_argument('--code', action='store_true')
parser.add_argument('--block', action='store_true')
opts = parser.parse_args()

def switch_to_branch_b_cb(uc, addr, *_):
    """
    When branch_A is called, switch to branch_B instead
    """
    print(f"PC: 0x{addr:08x}")

    if addr == BA_ADDR:
        print(f"Switching to branch B ({BB_ADDR:08x})")
        uc.reg_write(UC_ARM_REG_PC, BB_ADDR)


uc = unicorn.Uc(unicorn.UC_ARCH_ARM, unicorn.UC_MODE_ARM + unicorn.UC_MODE_LITTLE_ENDIAN)

uc.mem_map(CODE_ADDR, 1024)
uc.mem_write(CODE_ADDR, ARM_CODE)

uc.reg_write(UC_ARM_REG_PC, CODE_ADDR)
uc.reg_write(UC_ARM_REG_LR, EXIT_ADDR)

if opts.code:
    print("Using UC_HOOK_CODE")
    uc.hook_add(unicorn.UC_HOOK_CODE, switch_to_branch_b_cb)
if opts.block:
    print("Using UC_HOOK_BLOCK")
    uc.hook_add(unicorn.UC_HOOK_BLOCK, switch_to_branch_b_cb)

uc.emu_start(CODE_ADDR, EXIT_ADDR)

r0_value = uc.reg_read(UC_ARM_REG_R0)
print(f"Value of r0 after exec: 0x{r0_value:08x} (expecting {2:08x})")
```

**Comment by wtdcode** [comment]
Note when you register a `UC_HOOK_BLOCK`, it will only be triggered at 0x1000, 0x1008 and 0x100c in your case, which doesn't include `BA_ADDR` and the switch won't happen...

**Comment by oTuaz9gu** [comment]
Thanks for the reply! 

My bad, that's what I get for tinkering with my example and not being thorough enough reviewing it. I'm really sorry about that.

BA_ADDR was indeed set incorrectly in my code sample (+4 should have been +8). Once this is fixed, I continue to observe diverging behaviors between code and block hooks. 

When the PC is updated from the BLOCK hook, the CPU loops indefinitely on on the instruction at 0x1008 (first instruction of branchA). So it seems that the programmatic update to PC is ineffective, but the nominal update to PC (+4) also isn't happening either. 

I sincerely hope that the (hopefully) fixed example below doesn't contain any other silly mistakes.

---
```python
import argparse
import unicorn
from unicorn.arm_const import *

# main: 
#         mov r0, #0    ; 0x1000
#         b branch_A    ; 0x1004
#
# branch_A:
#         mov r0, #1    ; 0x1008
#         bx lr         ; 0x100c
#
# branch_B:
#         mov r0, #2    ; 0x1010
#         bx lr         ; 0x1014
ARM_CODE = bytes.fromhex('00 00 a0 e3 ff ff ff ea 01 00 a0 e3 1e ff 2f e1 02 00 a0 e3 1e ff 2f e1')

CODE_ADDR = 0x1000
EXIT_ADDR = 0x1100

BA_ADDR = CODE_ADDR + 0x8   # <- fixed!
BB_ADDR = CODE_ADDR + 0x10


parser = argparse.ArgumentParser()
parser.add_argument('--code', action='store_true')
parser.add_argument('--block', action='store_true')
opts = parser.parse_args()

def switch_to_branch_b_cb(uc, addr, *_):
    """
    When branch_A is called, switch to branch_B instead
    """
    print(f"PC: 0x{addr:08x}")

    if addr == BA_ADDR:
        print(f"Switching to branch B ({BB_ADDR:08x})")
        uc.reg_write(UC_ARM_REG_PC, BB_ADDR)


uc = unicorn.Uc(unicorn.UC_ARCH_ARM, unicorn.UC_MODE_ARM + unicorn.UC_MODE_LITTLE_ENDIAN)

uc.mem_map(CODE_ADDR, 1024)
uc.mem_write(CODE_ADDR, ARM_CODE)

uc.reg_write(UC_ARM_REG_PC, CODE_ADDR)
uc.reg_write(UC_ARM_REG_LR, EXIT_ADDR)

if opts.code or opts.block:
    expected = 2
else:
    expected = 1

if opts.code:
    print("Using UC_HOOK_CODE")
    uc.hook_add(unicorn.UC_HOOK_CODE, switch_to_branch_b_cb)
if opts.block:
    print("Using UC_HOOK_BLOCK")
    uc.hook_add(unicorn.UC_HOOK_BLOCK, switch_to_branch_b_cb)

uc.emu_start(CODE_ADDR, EXIT_ADDR)

r0_value = uc.reg_read(UC_ARM_REG_R0)
print(f"Value of r0 after exec: 0x{r0_value:08x} (expecting {expected:08x})")
```

**Comment by wtdcode** [comment]
It behaves good on my machine:

```
~/unicorn (dev|…) [2] $ python3 /tmp/test.py --block
Using UC_HOOK_BLOCK
PC: 0x00001000
PC: 0x00001008
Switching to branch B (00001010)
PC: 0x00001010
Value of r0 after exec: 0x00000002 (expecting 00000002)
~/unicorn (dev|…) $ python3 /tmp/test.py --code
Using UC_HOOK_CODE
PC: 0x00001000
PC: 0x00001004
PC: 0x00001008
Switching to branch B (00001010)
PC: 0x00001010
PC: 0x00001014
Value of r0 after exec: 0x00000002 (expecting 00000002)
~/unicorn (dev|…) $ python3 /tmp/test.py
Value of r0 after exec: 0x00000001 (expecting 00000001)
~/unicorn (dev|…) $
```

I assume it is some bug fixed already on dev branch?

**Comment by oTuaz9gu** [comment]
Confirmed, works fine on the `dev` branch. I gave up on bisecting due to compile time on my machine.

So to summarize: problem present 6c1cbef6ac505d355033aef1176b684d02e1eb3a and fixed between then and df3aa0fccbce9e1420e82110cbae5951755a0698

Thanks for looking into it.
